### ISSUE TITLE
Unconfirmed balance discrepancy

### ISSUE THREAD

**Issue by joostjager** [lightningnetwork/lnd#5685]
### Background

I am experimenting with two lnd nodes running alongside and initialized with the same seed. I want to find out to what extend a second lnd node can be a hot standby (onchain only). I ran into an issue that is also reproducible with just a single node.

### Your environment

* lnd v0.13.1-beta
* linux
* bitcoind

### Steps to reproduce

1. Set up an lnd node with a seed. I used:
 `abstract wheat arctic blade biology van this diamond radar economy trash blind trophy meadow draft bulk choose anger gain program adjust sword alone joke`

2. Generate a new address:
```
joost@joost-desktop $ lncli-alice newaddress p2wkh
{
    "address": "bcrt1qx4sqap808vkh4c065kcgavk590wdawxtlemrrf"
}
```

3. Delete all node data and recreate it with the same seed.

4. Send coin to the first wallet address:
```
bitcoin-cli sendtoaddress bcrt1qx4sqap808vkh4c065kcgavk590wdawxtlemrrf 0.1
```

5. Coin doesn't show up in wallet
```
lncli walletbalance
{
    "total_balance": "0",
    "confirmed_balance": "0",
    "unconfirmed_balance": "0",
    "account_balance": {
        "default": {
            "confirmed_balance": "0",
            "unconfirmed_balance": "0"
        }
    }
}
```

If you first run `newaddress` on the recreated instance and then send the coin, it is recognized.

### Expected behaviour

Node is always looking ahead and recognizing external wallet transactions.

Maybe this is a wrong assumption? If so, it does complicate the recovery flow. Addresses could have been generated before a node crash that haven't been used yet but will be used at some point.

### Actual behaviour

Wallet tx is not recognized.

**Comment by Roasbeef** [comment]
> Delete all node data and recreate it with the same seed.

If you set a recovery window during init, is the repro the same?

**Comment by joostjager** [comment]
Isn't that set to 2500 by default? I tried with 3000, but no difference.

Even if I mine the transaction created in step 4, it doesn't show up in the walletbalance.

**Comment by Roasbeef** [comment]
Perhaps it's a wallet birthday issue? Block height+time, etc on regtest/simnet can be weird. If for w/e reason, the birthday is such that we don't attempt to rescan, then I see how this can arise. Also note that we exercise this case in the itests (unless something is missing): https://github.com/lightningnetwork/lnd/blob/master/lntest/itest/lnd_recovery_test.go#L243-L277

**Comment by joostjager** [comment]
Yes, so I tried to explain above that this is not a recovery problem. I use recovery to reproduce with a single instance, but maybe that made it more confusing.

This is the two node repro:

1. Setup node A. Create a wallet and let lnd generate a fresh seed.
2. Setup node B. Create a wallet and use the seed that was generated by node A. Node B is now the 'hot standby'.
3. Generate a new address on node A.
4. Transfer coin to the new address.
5. Node A now shows the unconfirmed balance, while node B only shows zero amounts everywhere.
6. Mine a block
7. Node B still shows no balance.

Is this expected behavior?

**Comment by Roasbeef** [comment]
> Is this expected behavior?

If I'm following the scenario: yes, node B doesn't automatically re-create and derive the addresses that node A will use. Given there's no recovery situation here, there's no automatic look ahead implemented here. In order for those nodes to work in lock-step, a higher layer is needed to synchronize the wallet state.

> Generate a new address on node B. This will result in the same address that was obtained in step 3.

At this point, the address exists, but no historical rescan was done to find the transactions of the address, since from the PoV of node B it can't have been used here (the addr was just generated). If node A and B were to recovery from scratch, they'd have those same addresses, however as soon as some of them did a wallet state related action, they're de-synchronize again. Alternatively, if the addr created earlier were to _re used_ then B would pick that up.

**Comment by Roasbeef** [comment]
To take a peek under the hood a bit, this is the mthod that's used to read out all the active addresses that we need to watch from the wallet's PoV: https://github.com/btcsuite/btcwallet/blob/9b5a201c344c63be44705afafa974948e86b047b/wallet/wallet.go#L322-L351

That eventually reaches this method that scans for all the addresses _we've manually generated_: https://github.com/btcsuite/btcwallet/blob/9b5a201c344c63be44705afafa974948e86b047b/waddrmgr/db.go#L1831

Node B doesn't see those addresses yet, as that "new addr" call wasn't mirrored over. You'd see similar behavior if you manually used an xpub (or w/e) to derive the 10k-th address of a node, and sent funds there: if wouldn't see it until you did a new rescan, or manually inserted that new addr into the DB. 

We have code lying around (already used in the recovery scenario) to help remedy this, but our current approach is to do this at a higher level, given you need something that manages these set of nodes to ensure all the UTXOs, etc are in sync properly.

**Comment by joostjager** [comment]
>You'd see similar behavior if you manually used an xpub (or w/e) to derive the 10k-th address of a node, and sent funds there: if wouldn't see it until you did a new rescan, or manually inserted that new addr into the DB.

Reading through your explanation, I think this is not just the case for the 10k-th address that is used, but any address - even the first - that isn't derived through lnd? 

>our current approach is to do this at a higher level, given you need something that manages these set of nodes to ensure all the UTXOs, etc are in sync properly.

So how does this work in detail if one wants to have a mirror node standby? Make sure that every call to `NewAddress` is executed on both nodes? Maybe this isn't so easy to do because lnd is also deriving addresses internally. Also error cases may cause de-sync.

Also there is the scenario where a new address is generated and passed on to some external service that hasn't used the address yet. The node then crashes and is recovered with a rescan. The rescan doesn't see any trace of the address on-chain, because it hasn't been used yet. If the address is used thereafter by the external service, the transaction won't be picked up by lnd. Iirc the funds may never be recognized anymore unless a manual rescan is triggered, because once the block has passed it won't be re-examined.

How would you recommend to set up fail-over to an independent node with the same seed?

**Comment by Roasbeef** [comment]
> Reading through your explanation, I think this is not just the case for the 10k-th address that is used, but any address - even the first - that isn't derived through lnd?

Correct, the wallet as implemented is stateful: you either need to mirror all the operations, or pre-generate the addresses, or implement new logic that is _always_ maintaining a look head (and updates the state accordingly if a new transaction happens). 

> So how does this work in detail if one wants to have a mirror node standby? Make sure that every call to NewAddress is executed on both nodes? Maybe this isn't so easy to do because lnd is also deriving addresses internally. Also error cases may cause de-sync.

Yeah something like that, or something that does the state consistency checks and walks the keychains on either node forward when needed. Check out this PR for how mirroring would look like in theory: https://github.com/lightningnetwork/lnd/pull/5689

> The node then crashes and is recovered with a rescan. The rescan doesn't see any trace of the address on-chain, because it hasn't been used yet. If the address is used thereafter by the external service, the transaction won't be picked up by lnd. Iirc the funds may never be recognized anymore unless a manual rescan is triggered, because once the block has passed it won't be re-examined.

Well ideally the rescan sets an appropriate look ahead window (which will then always be extended and maintained when a used address is found). This addr state sync problem is pretty fundamental. Older versions of `bitcoind` just side stepped this by maintaining a large key pool when the wallet is first created (makes like 50k addresses or w/e). In the past, since not all wallets implement the same look ahead a gap window, when users would try to import a seed into another wallet, they would find their funds weren't showing up (unless the user manually created more addresses that necessary). 

> How would you recommend to set up fail-over to an independent node with the same seed?

If no changes to `btcwallet`, then you need something to always mirror any newly created or modified addrs across each node. If changes to `btcwalelt`, then one could take the existing rescan look ahead logic, and have that always apply when the wallet starts up, with the look ahead always being maintained each time new used addr is found (all this logic is already there, it just needs to be repurposed). 

If you take a look at that PR above, then in theory one could use the same method to create a "RAID wallet" by creating a new wrapper layer around lnd to replicate all addr state modifications to a cluster of `N` standby nodes. 

Alternatively, the `etcd` leader election (with everything in the main DB in master) can be used to do the very same thing (the state is already being replicated to all the `etcd` cluster nodes, and any of the newly elected `lnd` nodes will pick everything up). The same leader election code can also in theory be used with postgres as well, just more moving parts to manage.

**Comment by joostjager** [comment]
>Yeah something like that, or something that does the state consistency checks and walks the keychains on either node forward when needed. Check out this PR for how mirroring would look like in theory: #5689

Ok, yes that clarifies it. The de-sync risk is apparent in that code too. Also not sure how to go about syncing a new signer at a later time. The new signer would scan the chain, but there could be addresses that have been derived but not yet used.

> If you take a look at that PR above, then in theory one could use the same method to create a "RAID wallet" by creating a new wrapper layer around lnd to replicate all addr state modifications to a cluster of N standby nodes.

I get the idea, but I am not sure that this is the way to go, for the same reasons as mentioned above. What if one of the standby nodes is not available for address derivation?

It seems that the always-active look ahead makes for a more robust solution.

>Alternatively, the etcd leader election (with everything in the main DB in master) can be used to do the very same thing (the state is already being replicated to all the etcd cluster nodes, and any of the newly elected lnd nodes will pick everything up). The same leader election code can also in theory be used with postgres as well, just more moving parts to manage.

I don't think this is a true mirror. If one instance corrupts the database and a new leader is elected, that new leader will read into the very same corruption and there still won't be fail over.

It seems to me that on a virtualized platform like k8s, the lnd leader election feature isn't much different from spinning up a fresh lnd instance connected to the same database? Or am I missing the point here?
